### ISSUE TITLE
reordering a json may lead to "A value of known size was accessed past the end"

### ISSUE THREAD

**Issue by fuchsi-fuchs** [beached/daw_json_link#204]
There seems to be some issue with json_custom types - or I am missing something. I wanted to try daw_json for the public data from binance.com . It seemed to work fine at first - but when I implemented the custom type Fixed8 (replaced by `double` in the code below) the order of the fields `"a"` and `"b"` in the json suddenly seemed to make a difference. When the order in the json matches the oder in the json_data_contract declaration it works fine, if the two arrays are swapped an exception is thrown instead.

The order of keys in the input are not meant to make a difference, right?

```c++
const std::string testJson = R"({"e":"depthUpdate","E":1609884707320,"s":"BTCBUSD","U":2544556159,"u":2544556219,"a":[["34198.19000000","0.00000000"],["34198.23000000","0.00000000"],["34198.25000000","0.00000000"],["34198.27000000","0.00000000"],["34198.30000000","0.00000000"],["34198.32000000","0.00958500"],["34198.40000000","0.01232200"],["34198.41000000","0.01000000"],["34198.87000000","0.00000000"],["34199.12000000","0.00000000"],["34199.16000000","0.00000000"],["34199.42000000","0.00000000"],["34200.25000000","0.00000000"],["34200.71000000","0.03199900"],["34201.27000000","0.03100000"],["34201.62000000","0.00000000"],["34202.58000000","0.00000000"],["34204.45000000","0.00952700"],["34207.64000000","0.00000000"],["34207.74000000","0.00000000"],["34209.77000000","0.00000000"],["34209.81000000","0.20400000"],["34225.94000000","0.20200000"],["34226.60000000","0.91050000"],["34236.08000000","0.30000000"]],"b":[["34198.31000000","0.00000000"],["34196.54000000","0.00453200"],["34193.34000000","0.00000000"],["34189.89000000","0.00000000"],["34188.82000000","0.00000000"],["34185.32000000","0.00000000"],["34184.84000000","0.06350200"],["34184.83000000","0.20000000"],["34180.61000000","0.08622700"],["34180.60000000","0.00000000"],["34180.59000000","0.19200000"],["34180.02000000","0.00000000"],["34180.01000000","0.00000000"],["34176.88000000","0.00000000"],["34166.48000000","0.00000000"],["34166.47000000","0.00000000"],["34159.85000000","0.03317500"],["34159.24000000","0.09394900"],["34158.29000000","1.00000000"],["34154.86000000","0.00000000"]]})";

struct Fixed8JsonConverter {
	double operator( )( std::string_view sv ) const {
		return stod(std::string(sv));
	}
	
	template<typename OutputIterator>
	constexpr OutputIterator operator()(OutputIterator _it, double _f) const {
		return daw::json::utils::copy_to_iterator(_it, std::to_string(_f));
	}
};

template<JSONNAMETYPE name>
using json_fixed8 = daw::json::json_custom<name, double, Fixed8JsonConverter, Fixed8JsonConverter>;

struct Change {
	double rate;
	double amount;
};

namespace daw::json {
	template<>
	struct json_data_contract<Change> {
		using type = json_ordered_member_list<
						json_fixed8<no_name>, 
						json_fixed8<no_name>
					>;
	};
}

class DepthUpdateJson {
public:
	int64 time; // E
	std::string pairName; // s
	int64 idTo; // u
	int64 idFrom; // U
	std::vector<Change> bid; // b
	std::vector<Change> ask; // a
};

namespace daw::json {
	template<>
	struct json_data_contract<DepthUpdateJson> {
		using type = json_member_list<
			json_number<"E", int64>,
			json_string<"s">,
			json_number<"u", int64>,
			json_number<"U", int64>,
			json_array<"b", Change>,
			json_array<"a", Change>
		>;
	};
}

int main (const int argC, char* argV[] ) {
	try {
		auto parsed = daw::json::from_json<DepthUpdateJson>(testJson);
	} catch (daw::json::json_exception e) {
		std::cout << "daw error: " << e.reason() << " near: '" << e.parse_location() << "'"  << std::endl;
	}
}
```

The code produces the output
```
daw error: A value of known size was accessed past the end near: '["34198.23000000","0.00000000"],["34198.25000000","0.00000000"],["34198.27000000","0.00000000"],["34198.30000000","0.00000000"],["34198.32000000","0.00958500"],["34198.40000000","0.01232200"],["34198.41000000","0.01000000"],["34198.87000000","0.00000000"],["34199.12000000","0.00000000"],["34199.16000000","0.00000000"],["34199.42000000","0.00000000"],["34200.25000000","0.00000000"],["34200.71000000","0.03199900"],["34201.27000000","0.03100000"],["34201.62000000","0.00000000"],["34202.58000000","0.00000000"],["34204.45000000","0.00952700"],["34207.64000000","0.00000000"],["34207.74000000","0.00000000"],["34209.77000000","0.00000000"],["34209.81000000","0.20400000"],["34225.94000000","0.20200000"],["34226.60000000","0.91050000"],["34236.08000000","0.30000000"]],"b":[["34198.31000000","0.00000000"],["34196.54000000","0.00453200"],["34193.34000000","0.00000000"],["34189.89000000","0.00000000"],["34188.82000000","0.00000000"],["34185.32000000","0.00000000"],["34184.84000000","0.06350200"],["34184.83000000","0.20000000"],["34180.61000000","0.08622700"],["34180.60000000","0.00000000"],["34180.59000000","0.19200000"],["34180.02000000","0.00000000"],["34180.01000000","0.00000000"],["34176.88000000","0.00000000"],["34166.48000000","0.00000000"],["34166.47000000","0.00000000"],["34159.85000000","0.03317500"],["34159.24000000","0.09394900"],["34158.29000000","1.00000000"],["34154.86000000","0.00000000"]]}'
```

**Comment by beached** [comment]
Found where it is triggering in the error checking code. I'll look further into it tonight.  One question though, are you using strod because of the quotes in the double?  If so, the json_number parser supports three modes in one of it's flags, always quoted, sometimes quoted, and never quoted.  It looks like the two members of Change could be always quoted and not need the custom parser.

**Comment by fuchsi-fuchs** [comment]
Thanks for looking at it :-) I inserted the `stod` just to make this example as small as possible. In the actual code I am using a custom class to represent numbers with 8 decimal places and its from_string method instead of `double` and `stod`.

**Comment by beached** [comment]
ee2f429d4b62b1b204aecc34d284ed578621584a should fix/test this.  Just testing in CI with #205.

**Comment by beached** [comment]
The fix is now in release v2.7.2
